substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -1312,34 +1353,35 @@ where
 		);
 		let old_peers = std::mem::take(&mut self.peers);
 
-		old_peers.into_iter().for_each(|(peer_id, mut p)| {
-			// peers that were downloading justifications
-			// should be kept in that state.
-			if let PeerSyncState::DownloadingJustification(_) = p.state {
-				// We make sure our commmon number is at least something we have.
-				trace!(
-					target: LOG_TARGET,
-					"Keeping peer {} after restart, updating common number from={} => to={} (our best).",
-					peer_id,
-					p.common_number,
-					self.best_queued_number,
-				);
-				p.common_number = self.best_queued_number;
-				self.peers.insert(peer_id, p);
-				return
+		old_peers.into_iter().for_each(|(peer_id, mut peer_sync)| {
+			match peer_sync.state {
+				PeerSyncState::New | PeerSyncState::Available => {
+					self.add_peer(peer_id, peer_sync.best_hash, peer_sync.best_number);
+				},
+				PeerSyncState::AncestorSearch { .. } |
+				PeerSyncState::DownloadingNew(_) |
+				PeerSyncState::DownloadingStale(_) |
+				PeerSyncState::DownloadingGap(_) |
+				PeerSyncState::DownloadingState => {
+					self.add_peer(peer_id, peer_sync.best_hash, peer_sync.best_number);
+					self.actions.push(ChainSyncAction::CancelRequest { peer_id });
+					self.peer_pool.lock().free_peer(&peer_id);
```

Comment:
I wonder if here's a race condition. The peer is immediately freed but the cancellation is postponed until `SyncingEngine` processes the event so if some other strategy selects the peer while the request hasn't been canceled, it could result in two in-flight requests. Maybe the strategy/`SyncingStrategy` should be notified by `SyncingEngine` once the request is actually canceled so that freeing the peer is safe.

Reply:
It's even worse than that. When another strategy initiates the request, the old one is automatically dropped. And when we finally cancel the request, we can cancel a legitimate request of another strategy.

I'm not super happy with any kind of ACKing and bookkeeping, but it looks like we need to keep track of pending cancellations in every strategy and free peers in `PeerPool` when something like `on_request_cancelled()` is heard from `SyncingEngine`.

Reply:
Is there a reason why pending cancellations couldn't be stored in `SyncingStrategy` and once the request has been canceled, `SyncingEngine` calls `SyncingStrategy` which releases the peer?

Reply:
This means intercepting request cancellation actions in `SyncingStrategy` on their way to `SyncingEngine`. Not the best design ever, but this is probably better then dealing with it in every strategy.

Reply:
If we store peer's best and common blocks in `PeerPool`, we'd get rid of `peer_best_blocks`, allow `SyncingEngine` to replace its `HashMap<PeerId, Peer<B>>` with `PeerPool` and also allow it to free the peer once the request is canceled.

But now that I'm thinking this whole approach again, is there any fundamental reason why `GapSync` and `ChainSync` couldn't send a request to the same peer at the same time? I believe this limitation only applies to the sending end, namely the strategies, and made sense before when everything was one state machine. The request handlers shouldn't care how many requests from the same peer are in the queue. If `GapSync` and `ChainSync` are separate, is there a valid reason why they couldn't send simultaneous requests to a peer? We must ascertain that two copies of the same request are not sent because that'd get us banned, implying an existence of some kind of `PeerPool` for shared data, but is there a reason why they couldn't send two different requests at the same time? So the limitation of one request per peer still applies but now it's per strategy.

Reply:
> is there any fundamental reason why GapSync and ChainSync couldn't send a request to the same peer at the same time?

Valid point, there is nothing in the block request handler that forbids "simultaneous" requests. But if we get rid of the global `PeerPool`, we'll need to introduce a way to cancel specific requests, as otherwise strategies could cancel requests of each other.

Reply:
Each strategy could be identified by a unique key and when strategy returns `StrategyAction::SendRequest { PeerId, Request }`, `SyncingStrategy` would convert it to `SyncingAction::SendRequest { PeerId, Key, Request }` and `PendingResponses` would keep track of responses with `(PeerId, Key)`.

Reply:
Yes, that's what I had in mind.